src/libraries/System.Private.CoreLib/src/System/HashCode.cs

```diff
@@ -319,6 +320,38 @@ public void AddBytes(ReadOnlySpan<byte> value)
             ref byte pos = ref MemoryMarshal.GetReference(value);
             ref byte end = ref Unsafe.Add(ref pos, value.Length);
 
+            if ((nint)Unsafe.ByteOffset(ref pos, ref end) < sizeof(int) * 4)
+            {
+                goto Small;
+            }
+
+            // If we have more than 16 bytes to hash, we can add them in 16-byte batches,
+            // but we first have to add enough data to flush any queued values.
+            while (_length % 4 != 0 && (nint)Unsafe.ByteOffset(ref pos, ref end) >= sizeof(int))
```

Comment:
Some remarks about this loop:

* The second part of the `&&` will be always true thanks to the goto above (there are at least three 4-byte integers of input, and the length will always be incremented enough to be divisible by four, but I'm reluctant of removing it for fear of introducing out-of-bounds accesses in the future if this code gets somehow refactored, the invariant stops holding and nobody noticing it. Or I could just add a comment.
* We could even avoid the entire loop and directly modify the hash's state even if we have queued values. These will be processed later by either other calls to `Add` or calls to `ToHashCode` so no input is dropped, but I'm also reluctant because this is technically a divergence from the xxHash algorithm. It might not matter though, the underlying algorithm is an implementation detail, and it's not cryptographically sensitive either way.

Reply:
> the invariant stops holding and nobody noticing it. Or I could just add a comment.

Debug.Assert?

Reply:
👍, I think a `Debug.Assert` should be good enough to catch the refactoring's.

More generally, this is `1`, `2`, or `3` and loops do have a cost. It's probably better to make this `switch` with fallthrough instead so we have a single branch and then linear execution.